keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+- 2 single-family endpoints per backend pod: Make no change to the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core). Treat each IPv4/IPv6 address as separate, distinct endpoints, and include each address in the comma-separated list of addresses in an 'Endpoints' API object.
+- 1 dual-stack endpoint per backend pod: Modify the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) so that each endpoint can be associated with a pair of IPv4/IPv6 addresses.
+
+Although the first approach would be simpler and quicker to implement, providing two endpoints per backend pod would be problematic for the following ways in which endpoints are used in Kubernetes:
+- Distributed applications use endpoints for peer discovery.
+- Monitoring systems such as Prometheus use endpoints to identify monitoring targets.
+- Users use endpoints to determine how many instances are up.
+
+In order to avoid breaking the above uses of the Endpoints API, this design proposes modifying the [Kubernetes EndpointAddress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointaddress-v1-core) to support inclusion of both an IPv4 and an IPv6 address to be associated with a given endpoint.
+
+Note that this proposed change retains the singular sense of the API object name "EndpointAddress", even though the object may contain a pair of IPv4/IPv6 addresses. This can be considered a tradeoff in design simplification over clarity. The alternative would be to create a new API object called "EndpointAddresses" (similar to the "EndpointAddress" object, but with multiple IP addresses), and then modify the [Kubernetes EndpointSubset API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointsubset-v1-core) to include both the existing "EndpointAddress" object and a new "EndpointAddresses" object (for backwards compatibility).
+
+#### Versioned API Change: EndpointAddress v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal retains the existing (singular) "IP" field in the core V1 version of the [Kubernetes EndpointAddress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointaddress-v1-core), and adds a new slice of strings called "IPs" for recording multiple IPs associated with an endpoint:
+```
+// EndpointAddress is a tuple that describes one or more IP addresses for an
+// endpoint.
+ type EndpointAddress struct {
+     // The default IP for this endpoint.
+     IP string `json:"ip" protobuf:"bytes,1,opt,name=ip"`
+     // The IPs for this endpoint. The zeroth element (IPs[0] must match
+     // the default value set in the IP field)
+     IPs []string `json:"ips" protobuf:"bytes,5,opt,name=ips"`
```

Comment:
Do we want to have the same kind of "IP + metadata" that we have for PodIPs? Seems like that would be useful here too.

Reply:
the metadata option for PodIPs has a use case 
> refer to the "Properties" map below) will not be used by the dual-stack feature, but is added as a placeholder for future enhancements, e.g. to allow CNI network plugins to indicate to which physical network that an IP is associated

I can´t envision any use case for a metadata field in this case with IP endpoints but I think that we should add an example if we want to add it.

Reply:
@dcbw - do you have concrete use cases?  Remember that most endpoints objects are auto-created by the controller